### PR TITLE
Fix conversion Irrational => Rational

### LINKED ISSUE #16513: Rational{IntN}(pi) are poor approximations to Pi
These values are better approximations to Pi

Rational{Int32}(pi) = Int32(1068966896) // Int32(340262731)
Rational{Int64}(pi) = Int64(2646693125139304345) // Int64(842468587426513207)
Rational{Int128}(pi) = Int128(60728338969805745700507212595448411044) // Int128(19330430665609526556707216376512714945)

```
setprecision(BigFloat,500);
Q32pi =  Int32(1068966896) // Int32(340262731);
Q64pi = Int64(2646693125139304345) // Int64(842468587426513207);
Q128pi =  Int128(60728338969805745700507212595448411044) // Int128(19330430665609526556707216376512714945);

Float64(BigFloat(pi)-Q32), Float64(BigFloat(pi)-Rational{Int32}(pi)) # 3.1e-18, 1.2e-16
Float64(BigFloat(pi)-Q64), Float64(BigFloat(pi)-Rational{Int64}(pi)) # 1.4e-38, 1.2e-16
Float64(BigFloat(pi)-Q128) # -6.2e-76

### PR DESCRIPTION
**Pull request by simonbyrne**
Fixes #16513.

### CODE REVIEW

File: base/irrationals.jl [hunk 1]
```diff
@@ -13,7 +13,22 @@ promote_rule{s,T<:Number}(::Type{Irrational{s}}, ::Type{T}) = promote_type(Float
 convert(::Type{AbstractFloat}, x::Irrational) = Float64(x)
 convert(::Type{Float16}, x::Irrational) = Float16(Float32(x))
 convert{T<:Real}(::Type{Complex{T}}, x::Irrational) = convert(Complex{T}, convert(T,x))
-convert{T<:Integer}(::Type{Rational{T}}, x::Irrational) = convert(Rational{T}, Float64(x))
+
+@generated function convert{T<:Integer}(::Type{Rational{T}}, x::Irrational)
```
Comment (KristofferC) [inline base/irrationals.jl]: Maybe I am mistaken but this doesn't seem to return an expression like a `@generated` function should. It also doesn't really seem to need to be a `@generated` function at all.

Comment (simonbyrne) [inline base/irrationals.jl]: Returning a value is okay (it just gets spliced in as an expression). 

The aim of the `@generated` is to get it to perform the calculation at compile time, as the result will always be constant. This is exactly the same as the usage in the function directly below.

Technically, an `@pure` annotation would be sufficient , once #14324 is sorted out.

File: base/irrationals.jl [hunk 2]
```diff
@@ -13,7 +13,22 @@ promote_rule{s,T<:Number}(::Type{Irrational{s}}, ::Type{T}) = promote_type(Float
...
+    o = precision(BigFloat)
+    p = 256
+    while true
+        setprecision(BigFloat, p)
+        bx = BigFloat(x())
```
Comment (tkelman) [inline base/irrationals.jl]: how does this work? I get
```
julia> pi()
ERROR: MethodError: objects of type Irrational{:π} are not callable
```

Comment (hayd) [inline base/irrationals.jl]: It seems that `x == typeof(pi)`:

```
julia> @generated function foo(x) x end
foo (generic function with 1 method)

julia> foo(pi)
Irrational{:π}

julia> foo(pi)()
π = 3.1415926535897...
```
Is that a bug?

Comment (tkelman) [inline base/irrationals.jl]: right, no, this is a generated function, this is how they're supposed to work which I just keep forgetting

**Comment by TotalVerb** [comment]
Going to crosspost this from #16513.

Is this operation meaningful? Non-exact operations on `Rational`s either return `Float64` or throw `OverflowError`. This would be a departure from that.

Also, converting from an `Irrational` to a `Rational` seems weird in principle.

**Comment by ivarne** [comment]
@TotalVerb This pr does not depart from anything. This only fixes a roundoff error (on a dubious conversion).

It would be a different issue to just remove the conversion (with a `InexactError`, or `MethodError`), but I'd be against that because `Irrational` isn't a computation type.

@simonbyrne Given the confusions about `@generated` it would be nice to have a comment above the method explaining why it is written as it is written. 

(Maybe something like)
```julia
# This is a @generated function because it allows the actual calculation to be performed compile-time 
# Note that it returns a Rational constant that get's spliced into the AST of the caller
```

**Comment by simonbyrne** [comment]
At the moment, the [`convert` docs](http://docs.julialang.org/en/latest/stdlib/base/#Base.convert) say:

>   If `T` is a `AbstractFloat` or `Rational` type, then it will return the closest value to `x` representable by `T`.

File: base/irrationals.jl (hunk 1 above)
Comment (JeffBezanson) [inline base/irrationals.jl]: I think putting `@pure` on this function would just work.

Comment (simonbyrne) [inline base/irrationals.jl]: I tried, it didn't.

Comment (simonbyrne) [inline base/irrationals.jl]: See https://gist.github.com/simonbyrne/026a6969a3316cf04aa95c5c75d9dacd

**Comment by simonbyrne** [comment]
Using `@generated` was vetoed, so this has been changed to `@pure`.

Should be ready to merge.

**Pull request merged**